core/consensus/grandpa/impl/grandpa_impl.cpp

```diff
@@ -1169,45 +1247,80 @@ namespace kagome::consensus::grandpa {
-            peer_id, network::reputation::benefit::BASIC_VALIDATED_COMMIT);
-        return;
-      }
-
-      if (ctx->missing_blocks.empty()) {
-        SL_WARN(logger_,
-                "Commit with set_id={} in round={} for block {} "
-                "has received from {} and has not applied: {}",
-                msg.set_id,
-                msg.round,
-                BlockInfo(msg.message.target_number, msg.message.target_hash),
-                peer_id,
-                res.error());
-        return;
-      }
+      applyJustification(
+          justification,
+          [wself{weak_from_this()},
+           check_missed_blocks{std::move(check_missed_blocks)},
+           peer_id,
+           msg,
+           grandpa_context{std::move(grandpa_context)}](auto &&res) mutable {
+            if (auto self = wself.lock()) {
+              if (res.has_value()) {
+                self->reputation_repository_->change(
+                    peer_id,
+                    network::reputation::benefit::BASIC_VALIDATED_COMMIT);
+                return;
+              }
+
+              if (grandpa_context.missing_blocks.empty()) {
+                SL_WARN(self->logger_,
+                        "Commit with set_id={} in round={} for block {} "
+                        "has received from {} and has not applied: {}",
+                        msg.set_id,
+                        msg.round,
+                        BlockInfo(msg.message.target_number,
+                                  msg.message.target_hash),
+                        peer_id,
+                        res.error());
+                return;
+              }
+              check_missed_blocks(std::move(grandpa_context));
+            }
+          });
     } else {
-      ctx->missing_blocks.emplace(justification.block_info);
+      grandpa_context.missing_blocks.emplace(justification.block_info);
+      check_missed_blocks(std::move(grandpa_context));
     }
+  }
 
-    // Check if missed block are detected and if this is first attempt
-    // (considering by definition peer id in context)
-    if (not ctx->missing_blocks.empty()) {
-      if (not ctx->peer_id.has_value()) {
-        ctx->peer_id.emplace(peer_id);
-        ctx->commit.emplace(msg);
-        loadMissingBlocks();
-      }
-    }
+  void GrandpaImpl::callbackCall(ApplyJustificationCb &&callback,
+                                 outcome::result<void> &&result) {
+    main_thread_context_.execute(
+        [callback{std::move(callback)}, result{std::move(result)}]() mutable {
+          callback(std::move(result));
+        });
   }
 
-  outcome::result<void> GrandpaImpl::verifyJustification(
-      const GrandpaJustification &justification,
-      const primitives::AuthoritySet &authorities) {
+  void GrandpaImpl::verifyJustification(const GrandpaJustification &jst,
+                                        const primitives::AuthoritySet &auth,
+                                        ApplyJustificationCb &&cb) {
```

Comment:
`WarpSync` verifies multiple justifications in for-loop.
Would it be better to remove callback from `verifyJustification` and run `WarpSync` for-loop on other thread?